Workflow Editor phase 2, part 2: Step UI on Task funnel

shaunanoordin opened the pull request:
## PR Overview

Part of: **Workflow Editor** (aka **Pages Editor**) project phase 2
Related issue: #7241
Follows #7245
Staging branch URL: https://pr-7253.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging

This PR is the second UI update for the Workflow Editor project, phase 2. This time, we're updating (in designer speak:) the Task UI elements on the Task funnel, aka (in dev speak:) the StepItem components on the Tasks page.

### Status

~~WIP~~

Update: [Ready for review!](https://github.com/zooniverse/Panoptes-Front-End/pull/7253#issuecomment-2672956605)

seanmiller26 commented:
@shaunanoordin I recently made some final adjustments to the order of the Edit, Copy, and Delete buttons. This updated arrangement follows a more logical sequence and improves usability, particularly for tab navigation. Hover and click states were also updated for platform consistency.

Please see #7241 and the Figma file for details.

shaunanoordin commented:
## PR Update

This PR is now ready for review! See #7241 for details, but here's the overview:-

Quick guide to terminology:

<img width="600" alt="image" src="https://github.com/user-attachments/assets/e6336053-5034-4aed-ac61-d7213bbd2ee1" />

<img width="600" alt="image" src="https://github.com/user-attachments/assets/da34e2cf-5acf-4e54-83e3-af734c2ff429" />

Design updates applied to the Tasks page:[^1]
- Pages have been restyled.
  - Pages now display their Page labels, e.g. "Page 1".[^2] Note that a Page label is distinct from Page ID/Step Key, e.g. "P0".
  - Page control icons (delete page, reorder page up/down, edit page, etc) have been updated/replaced. 
- Question Task (single & multiple), Drawing Task, and Text Task have been restyled.
  - Tasks should appear closer in visual design to what we see on the FEM Classifier. Hopefully.
  - Each Task now has a unique Task Icon.
- Custom **Task Icons** added. These are subject to change, and for now use the PNG file format. (See Slack [[1]](https://zooniverse.slack.com/archives/C06LHUC6R/p1739989891236549?thread_ts=1739977229.852399&cid=C06LHUC6R)[[2]](https://zooniverse.slack.com/archives/C06LHUC6R/p1740077041848649) for a larger conversation about custom Zooniverse icons)

Also:
- README: added notes on a "Future migration guide to FEM"

[^1]: the "Edit Task/Edit Page" dialogs haven't been touched
[^2]: Page label numbers are based on the _index_ of the Page. (i.e. the nominal order that volunteers will experience the workflow, barring branching.) So if you have `[P0, P1, P2] => Page 1, Page 2, Page 3`, then rearrange it, you'll still get `[P2, P0, P1] => Page 1, Page 2, Page 3`

### Testing

To recap, this is primarily a UI update. @seanmiller26 , please continue reviewing and providing feedback as you already have. 👌 

For devs, your testing goals are to ensure that...
- as a _project owner,_ you should be able to create and edit Pages and Tasks without any UI-related confusion.
- you should be able to understand the workflow you've created, just by looking at the Pages on the _Tasks page_ (aka "Task funnel")
- you should be able to view & edit the workflow on most modern browsers.
- (accessibility) you should be able to perform the work you want (add, edit, delete, rearrange, etc) with either keyboard or mouse.
- (accessibility) the information on the page should still be understandable if you're vision-impaired.

Here's the recommended testing method:
- View a test workflow on staging, e.g. https://pr-7253.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging
- Create different Task types and Page combinations, e.g.:
  - create a Page with 1 Drawing Task with 3 different Drawing Tools
  - create a Page with 3 Text Tasks
  - create a Page with 1 Question Task with 3 answers, each branching to a different Next Page or Submit.
- Test for cases that could theoretically break the design, e.g.:
  - Pages/Steps with _too many_ Tasks. [(see Slack)](https://zooniverse.slack.com/archives/C39AUB9UK/p1739920007834579?thread_ts=1739919640.296939&cid=C39AUB9UK)
  - Different browsers.[^4]

[^3]: OK, so _one_ known instance of visual shenanigans that happens when a Page has too many tasks is that ALL the "Next Page" select dropdowns appear very long/too wide, because they'll list things like "T1, T2, T3, T4, T5, T6, T7, T8". There's discussion on Slack about truncating those things to "T1, T2, T3... (dot dot dot)".
[^4]: The "Next Page" select - and every `<select>` element - looks terrible in Safari.

### Status

Ready for review!

seanmiller26 commented:
- [ ] Page padding should be standardized to 20px, 30px on bottom (except Question tasks, set to 20px all sides. See below)
- [ ] Header/step-controls set to 45px height, remove margin
- [ ] Reduce paragraph font weight
- [ ] Task key should be standardized to 20px x 30px
- [ ] Step label should have 5% letter spacing
- [ ] Replace all instances of 'contributor' with 'volunteer'

Drawing tool specific:
- [ ] The drawing tools should be 60px height

Safari specific:
I don't mind Safari's default styling for the dropdowns, but we'll still want to address page padding.

<img width="200" alt="image" src="https://github.com/user-attachments/assets/31f5a7e6-af28-42b2-8964-51c079e81f8c" />

All task padding:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/1b489dc0-4bbb-4529-909e-5e9776912f21" />

Question task padding:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/475ce8c6-9900-412b-a8d6-7662072a7c78" />



Questions: 

Should the Edit, Duplicate, and Delete buttons be labelled more descriptively than svg.icon ? If this is being addressed elsewhere, like tabbing, aria labels, etc, then disregard.

We may want to rethink the label of 'fake' buttons. I'm not tied to a specific phrase, but I'm more comfortable with placeholder, simulated or mock than fake.

_Probably_ not an ideal mobile view, will need to revisit:
<img width="400" alt="Screenshot 2025-02-27 at 12 38 24 PM" src="https://github.com/user-attachments/assets/4795a584-fc27-4a33-9824-beb865eeaca8" />

Early concept which may come in handy:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/fd54c139-8e62-40ec-8dee-361be1db899f" />


All details available in Figma: https://www.figma.com/design/gNOcFdlskfEXmVIFPvx7Yj/Workflow-Editor?node-id=1484-10575&m=dev

goplayoutside3 reviewed (approved):
This looks great! Sean's already provided design feedback so I only have a couple of things to add to his review.

1. Re:
> Should the Edit, Duplicate, and Delete buttons be labelled more descriptively than svg.icon ? If this is being addressed elsewhere, like tabbing, aria labels, etc, then disregard.

- These have aria-labels on the `<button>` elements. However, I did notice the newest icons that sit next to each task name (T0, T1, etc) are `<img>` elements missing alt tags. If you want them to be ignored by screen readers, add `alt=""`, else a descriptive alt tag to be intentionally read.

2. Re:
> We may want to rethink the label of 'fake' buttons. I'm not tied to a specific phrase, but I'm more comfortable with placeholder, simulated or mock than fake.

- 🤔  What part of the UI is this referring to?

3. A couple of layout notes non-specific to this PR:
    - According to Figma (and also my opinion) the new workflow editor's header shouldn't have any space in between it and the project builder's nav on the left. It's most noticeable how the teal header doesn't sit flush against the nav. This is just a suggestion to add a style of `margin-left: -1.5vw` to `.lab-pages-editor` which will offset the `<hr />` that sits in between `<nav class="Lab navigation">` and `<main/>` in project.jsx of the lab. Kind of hacky 🤷‍♀️ but minimal effects on legacy lab content.
    - I also noticed there's no space to the left of the `<nav class="Lab Navigation>"` specifically when viewing the PagesEditor tab. All other project builder tabs have a margin-left of `3vw`. There's also margin around the whole `class="columns-container"` but I see why you got rid of this in order for the PagesEditor header to sit next to the zoo header. Again, just a suggestion to add `margin-left: 3vw` to the `.pages-editor-container` to match the rest of the lab. Here's a screenshot of both layout notes:
    
<img width="363" alt="layout of side nav plus .pages-editor-container" src="https://github.com/user-attachments/assets/45675021-d9a4-4eba-b4c7-4ac951c66c0a" />

shaunanoordin commented:
Thanks Sean & Delilah for the feedback and review! I've addressed some of the items discussed, but primarily I'll go ahead and merge this PR first and work on improvements in the next few PRs.

--------

**For Sean:**

I've attempted to address all the low-hanging items, including: 1. Contributor -> Volunteer, 2. Step Labels have 5% letter spacing, 3. paragraph of Task main text set to font-weight 500, 4. all "fake" elements renamed to "mock" elements.

Everything else - notably the layout/size stuff that requires more thought on my end - is on the TODO, and I've opened #7276 to make sure I don't forget to address these things.

Also, _when a branching Question Task is the last Task on a page,_ and _the window width is narrow,_ the Answers are supposed to display like this. (screenshot below) But there's a bug where this layout doesn't correct itself _when the window is resized,_ so you see the overlapping view you screenshotted. Boo! Flagging this for a discussion/re-discussion(?) on a call to get the correct behaviour, as there are a few permutations I need to walk through. 🎏 

<img width="337" alt="image" src="https://github.com/user-attachments/assets/94f19106-b25f-42e2-a1d5-a1ac68b2e710" />

--------

**For Delilah**

I've added your layout suggestions (item 3) to 7276.

As for the "fake buttons" (now "mock buttons") that Sean was talking about, he's referring to the UI elements _that look like inputs and buttons, but are actually not interactive, and are only used to simulate interactive elements the volunteer would see on the classifier._

In the screenshot below, the "volunteer text here", "foo", and "bar" elements look like an input.type=text, button, and button respectively. However, they're actually _divs_ styled to look like buttons and text input.

<img width="405" alt="image" src="https://github.com/user-attachments/assets/c3370607-0a18-43d8-8e38-f48a7b8d7727" />

--------

**Accessibility notes:**

I've marked which elements have additional accessibility in the code. In the screenshots below, all buttons with aria labels (CSS selector: `button[aria-label]`) are marked in red. All images with alt text (CSS selector: `img[alt]`) are marked in green. Notably, the ~~fake~~ mock buttons 

<img width="400" alt="image" src="https://github.com/user-attachments/assets/e3601222-4222-42d7-bdbf-634e45909505" />

<img width="400" alt="image" src="https://github.com/user-attachments/assets/1990fd34-6030-4685-bf0a-4e8fdd79e628" />

This definitely deserves another pass once more icons & controls are updated.

--------

Anyway, let's merge and go! Thanks again, both!
